Mathlib/Probability/CondJensen.lean

```diff
@@ -0,0 +1,577 @@
+              _ = infDist x s := by simp only [add_halves]
+          exact notMem_of_dist_lt_infDist hxy hy1
+        have hxfi : x ∈ (ball ↑(f i) (infDist ↑(f i) s)) := by
+          rw [mem_ball]
+          calc
+            dist x ↑(f i) < (infDist x s) / 2 := hi
+            _ ≤ infDist ↑(f i) s := hfix
+        exists i; apply (hLc i).1; exact hxfi
+      · intro hx i
+        exact ((hLc i).2 x hx)
+    · intro ha j
+      cases le_or_gt (c j) 0 with
+      | inl hl =>
+        use f j; apply ne_of_lt; apply lt_of_lt_of_le
+        · have : (⇑re ∘ ⇑(L j)) ↑(f j) < c j := by
+            apply (hLc j).1; apply mem_ball_self
+            have : ↑(f j) ∈ sᶜ := (f j).property
+            exact (IsClosed.notMem_iff_infDist_pos hs₂ hs₃).mp this
+          exact this
+        · exact hl
+      | inr hr =>
+        simp only [nonempty_def] at hs₃
+        rcases hs₃ with ⟨x, hxs⟩; use x
+        apply ne_of_gt; apply lt_of_lt_of_le
+        · exact hr
+        · have : c j ≤ (⇑re ∘ ⇑(L j)) x := by
+            apply (hLc j).2; exact hxs
+          exact this
+
+/-- Lemma 1.2.9 for product spaces. -/
+theorem iInter_halfSpaces_eq_of_separableSpace_prod {𝕜 E F : Type*} {s : Set (E × F)}
+    [RCLike 𝕜] [NormedAddCommGroup E] [NormedSpace ℝ E]
+    [SeparableSpace E] [Module 𝕜 E] [ContinuousSMul 𝕜 E]
+    [NormedAddCommGroup F] [NormedSpace ℝ F]
+    [SeparableSpace F] [Module 𝕜 F] [ContinuousSMul 𝕜 F]
+    (hs₁ : Convex ℝ s) (hs₂ : IsClosed s) (hs₃ : s.Nonempty) :
+    ∃ (L : ℕ → E →L[𝕜] 𝕜) (T : ℕ → F →L[𝕜] 𝕜) (c : ℕ → ℝ),
+    ⋂ (i : ℕ), {(x, y) | re ((L i) x) + re ((T i) y) - c i ≥ 0} = s
+    ∧ (sᶜ.Nonempty → ∀ i, ∃ (x : E), ∃ (y : F), (re ∘ L i) x + (re ∘ T i) y ≠ 0) := by
+  have lem1 : ∃ (LT : ℕ → (E × F) →L[𝕜] 𝕜) (c : ℕ → ℝ), ⋂ (i : ℕ), {x | re ((LT i) x) - c i ≥ 0} = s
+    ∧ (sᶜ.Nonempty → ∀ i, ∃ x, (re ∘ LT i) x ≠ 0) :=
+    iInter_halfSpaces_eq_of_separableSpace hs₁ hs₂ hs₃
+  rcases lem1 with ⟨LT, c, eq1, eq2⟩
+  exists fun i ↦ ((LT i).comp (inl 𝕜 E F))
+  exists fun i ↦ ((LT i).comp (inr 𝕜 E F))
+  exists c
+  have lem2 : ∀ (x : E), ∀ (y : F), (x, y) = (x, 0) + (0, y) := by
+    intro x y; simp only [Prod.mk_add_mk, add_zero, zero_add]
+  have lem3 : ∀ i, ∀ (x : E), ∀ (y : F), re ((LT i) (x, 0)) + re ((LT i) (0, y))
+     = re ((LT i) (x, y)) := by
+    intro i x y; simp only [lem2 x y, (LT i).map_add, re.map_add]
+  constructor
+  · rw [← eq1]; apply iInter_congr; intro i; ext ⟨x, y⟩
+    simp only [coe_comp', Function.comp_apply, inl_apply, inr_apply, ge_iff_le, sub_nonneg,
+      mem_setOf_eq]
+    rw [lem3]
+  · intro hsc i; rcases eq2 hsc i with ⟨z, hz⟩
+    use z.1; use z.2; simp only [coe_comp', Function.comp_apply, inl_apply, inr_apply, lem3]
+    exact hz
+
+end RCLike
+
+/-- Lemma 1.2.10 in [Hytonen_VanNeerven_Veraar_Wies_2016]: a convex lower-semicontinuous function
+  is the supremum of a sequence of affine functions in a separable space. -/
+theorem ConvexOn.iSup_affine_eq_of_separableSpace {𝕜 E : Type*}
```

Comment:
That lemma is about convex functions in general and not specifically about conditional expectations, so it should be moved to a file about convex functions.

All the auxiliary lemmas of this file should be moved to other more appropriate files.